src/random_lcg.C

```diff
@@ -0,0 +1,142 @@
+#include "random_lcg.h"
+#include <cmath>
+
+
+// Starting seed
+int64_t seed = 1;
+
+// LCG parameters
+const uint64_t prn_mult   = 2806196910506780709LL;   // multiplication factor, g
+const uint64_t prn_add    = 1;                       // additive factor, c
+const uint64_t prn_mod    = 0x8000000000000000;      // 2^63
+const uint64_t prn_mask   = 0x7fffffffffffffff;      // 2^63 - 1
+const uint64_t prn_stride = 152917LL;                // stride between particles
+const double   prn_norm   = pow(2, -63);             // 2^-63
+
+// Module constants
+const int N_STREAMS         = 5;
+const int STREAM_TRACKING   = 0;
+const int STREAM_TALLIES    = 1;
+const int STREAM_SOURCE     = 2;
+const int STREAM_URR_PTABLE = 3;
+const int STREAM_VOLUME     = 4;
+
+// Current PRNG state
+uint64_t prn_seed[N_STREAMS];  // current seed
+int      stream;               // current RNG stream
+#pragma omp threadprivate(prn_seed, stream)
+
+
+//==============================================================================
+// PRN
+//==============================================================================
+
+extern "C" double
+prn()
+{
+  // This algorithm uses bit-masking to find the next integer(8) value to be
+  // used to calculate the random number.
+  prn_seed[stream] = (prn_mult*prn_seed[stream] + prn_add) & prn_mask;
+
+  // Once the integer is calculated, we just need to divide by 2**m,
+  // represented here as multiplying by a pre-calculated factor
+  return prn_seed[stream] * prn_norm;
+}
+
+//==============================================================================
+// FUTURE_PRN
+//==============================================================================
+
+extern "C" double
+future_prn(uint64_t n)
+{
+  return future_seed(n, prn_seed[stream]) * prn_norm;
+}
+
+//==============================================================================
+// SET_PARTICLE_SEED
+//==============================================================================
+
+extern "C" void
+set_particle_seed(uint64_t id)
+{
+  for (int i = 0; i < N_STREAMS; i++) {
+    prn_seed[i] = future_seed(id * prn_stride, seed + i);
+  }
+}
+
+//==============================================================================
+// ADVANCE_PRN_SEED
+//==============================================================================
+
+extern "C" void
+advance_prn_seed(uint64_t n)
+{
+  prn_seed[stream] = future_seed(n, prn_seed[stream]);
+}
+
+//==============================================================================
+// FUTURE_SEED
+//==============================================================================
+
+extern "C" uint64_t
+future_seed(uint64_t n, uint64_t seed)
```

Comment:
So one thought is that we should just have the signature on the C++ side match what the interfaces defined on the Fortran side say, so `n` and `seed` would both be `int64_t` (this applies to the other functions as well). Conversion between signed and unsigned integers in C and C++ *is* well-defined, so long as that conversion happens explicitly in the body of the function rather than implicitly by passing an argument and relying on the similar bit representation. Citing section 7.8, point 2 in the [C++17 draft standard](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/n4700.pdf):
> If the destination type is unsigned, the resulting value is the least unsigned integer congruent to the source integer (modulo 2 n where n is the number of bits used to represent the unsigned type). [ Note: In a two’s complement representation, this conversion is conceptual and there is no change in the bit pattern (if there is no truncation). — end note ]

Reply:
Good idea.  I've replaced all the arguments that talk to Fortran with `int64_t`s.